Pre-request code actions

tomv564 opened the pull request:
The Hover popup currently always shows a Code Actions link, even if no code actions are available. This PR aims to query the server, then show information about available code actions in the popup.

This shows an initial implementation, running with a typescript and eslint server.
[
<img width="639" alt="Screenshot 2019-10-01 at 11 30 23" src="https://user-images.githubusercontent.com/4395832/65950717-e99d0d80-e43e-11e9-8bbd-29eaf9b55bfc.png">
](url)

**Two functional issues need to be resolved:**
- [x] The list of code actions is long - change to a "Code Actions (3)" link?
- [x] Only the first codeActionProvider for the document is being queried - we should query all and show a "Code Actions" link for each.
  - (I'm hesitant to get into this, but of all places to combine LSP results, this use case probably makes the most sense)

**UI opinions wanted**

The hover UI is getting messy. Code actions feel like they "belong" to the diagnostics, but there can be many blocks of diagnostics (errors and warnings, per server) so it's unclear where to place the links.

**Technical fixes**
- [x] Potentially cooperating with the bulb logic (so fetched actions can be used at once)

tomv564 commented:
Last commit achieved the two feature goals: All code action features to query all servers, and the hover shows an actions link per server diagnostics section.

**Examples**

Both servers report a diagnostic, only typescript offers actions:

<img width="569" alt="Screenshot 2019-10-03 at 13 04 55" src="https://user-images.githubusercontent.com/4395832/66122020-ee98c300-e5de-11e9-8dc5-a742a91cf65c.png">

Eslint with error, warning and actions block:

<img width="433" alt="Screenshot 2019-10-03 at 13 05 08" src="https://user-images.githubusercontent.com/4395832/66122030-f5273a80-e5de-11e9-9d44-a0fa66559ec1.png">

**UI polish goals (please suggest how you'd improve this!)**
* Reduce the amount of color in diagnostics
* Get the Code Actions link to pop out a bit more (is it rendering darker than the diagnostics texts?)

jwortmann commented:
I just tested this quickly:
* The logic to not show a 'Code Actions' link if the server doesn't provide it seems to work fine
* Something isn't right with the popup HTML/CSS yet, because the `warnings` class is omitted for me now:
Before:
![before](https://user-images.githubusercontent.com/6579999/66125089-60283f80-e5e6-11e9-884f-223dda30f93e.png)
After:
![after](https://user-images.githubusercontent.com/6579999/66125101-68807a80-e5e6-11e9-88fb-c3438728f401.png)

My opinion for the UI:
I would prefer to not have the background color `--accent` for the 'Code Actions' link CSS, because it could be an arbitrary color from the color scheme, which might even be the same as `--yellowish` or `--redish` for warnings and errors. In your case it is blue, but even that doesn't seem optimal if the server provides a info/hint message, which has also blue background. I think the traditional underlined blue text link without background color (like in your opening comment) would be better here, so that users immediately see that it is a clickable link.

tomv564 commented:
Thanks for the feedback! I added the missing class and adjusted the colors - I hear your argument for the underlined link, but I'd like to try a button-like style as well to differentiate LSP actions from documentation links.

<img width="405" alt="Screenshot 2019-10-03 at 16 09 18" src="https://user-images.githubusercontent.com/4395832/66134364-848d1780-e5f8-11e9-9a68-c305522d8ef4.png">

jwortmann commented on plugin/hover.py:
```diff
@@ -126,28 +131,38 @@ def symbol_actions_content(self) -> str:
+                                                            diagnostic_message)
         else:
-            return "<pre>{}</pre>".format(diagnostic_message)
+            return "<pre class=\"{}\">{}</pre>".format(diagnostic_message)
```
```suggestion
            return "<pre class=\"{}\">{}</pre>".format(class_for_severity[diagnostic.severity], diagnostic_message)
```

jwortmann commented on plugin/hover.py:
```diff
@@ -173,7 +188,15 @@ def hover_content(self, point: int, response: 'Optional[Any]') -> str:
+        _test_contents.clear()
+        _test_contents.append(contents)  # for testing only
+
         mdpopups.show_popup(
```
I think there should be
```
if contents:
    mdpopups.show_popup(
    ...
    ...)
```
to prevent empty popups.

rchl commented:
I've tried it out just for a minute and one thing that bothered me immediately was that the hit zone for the `Core Actions` link is only on the characters and not on the spaces. Very unline normal links on web pages. :)

tomv564 commented:
Reverted the Code Actions link to the classic blue underlined after your feedback @jwortmann  and @rchl !
